File: python/hush-icore/hush/core/ops/transform/parser_op.py
"""ParserOp — extract structured data from text."""

import json
import xml.etree.ElementTree as ET
from dataclasses import dataclass
from typing import Any, Dict, List, Literal, Optional

import yaml

from hush.core.configs.op_config import OpType
from hush.core.exceptions import ParserError
from hush.core.loggings import LOGGER
from hush.core.ops.base import BaseOp
from hush.core.utils.common import Param

ParserType = Literal["json", "xml", "yaml"]


@dataclass
class ExtractField:
    """Biểu diễn một field cần trích xuất với path và thông tin type."""

    output_key: str
    chain_path: List[str]
    type_hint: str

    @classmethod
    def from_string(cls, schema_str: str) -> "ExtractField":
        """Parse chuỗi schema như 'company.user.address: dict' thành ExtractField."""
        if ":" not in schema_str:
            schema_str += ": Any"

        chain_text, type_hint = schema_str.split(":", 1)
        chain_path = chain_text.strip().split(".")
        output_key = chain_path[-1]

        return cls(output_key=output_key, chain_path=chain_path, type_hint=type_hint.strip())


def parse_json(text: str) -> Dict[str, Any]:
    """Parse JSON text."""
    text = text.strip()
    if text.startswith("```"):
        lines = text.split("\n")
        text = "\n".join(lines[1:-1]) if len(lines) > 2 else text
    return json.loads(text)


def parse_xml(text: str) -> Dict[str, Any]:
    """Parse XML text thành dictionary.

    Handles both single-root and multiple top-level elements.
    For multiple elements like <a>1</a><b>2</b>, wraps in <root> and flattens.
    """

    def xml_to_dict(element):
        result = {}
        for child in element:
            if len(child) == 0:
                result[child.tag] = child.text
            else:
                child_dict = xml_to_dict(child)
                if child.tag in result:
                    if not isinstance(result[child.tag], list):
                        result[child.tag] = [result[child.tag]]
                    result[child.tag].append(child_dict)
                else:
                    result[child.tag] = child_dict
        return result

    text = text.strip()
    if text.startswith("```"):
        lines = text.split("\n")
        text = "\n".join(lines[1:-1]) if len(lines) > 2 else text

    # Try parsing as-is first
    try:
        root = ET.fromstring(text)
        return {root.tag: xml_to_dict(root)} if len(root) > 0 else {root.tag: root.text}
    except ET.ParseError:
        # Multiple root elements - wrap in <root> and flatten result
        wrapped = f"<root>{text}</root>"
        root = ET.fromstring(wrapped)
        return xml_to_dict(root)


def parse_yaml(text: str) -> Dict[str, Any]:
    """Parse YAML text."""
    try:
        text = text.strip()
        if text.startswith("```"):
            lines = text.split("\n")
            text = "\n".join(lines[1:-1]) if len(lines) > 2 else text
        return yaml.safe_load(text)
    except ImportError:
        raise ImportError("pyyaml là bắt buộc để parse YAML")


# Module-level parser lookup for O(1) format selection
_PARSER_MAP = {
    "json": parse_json,
    "xml": parse_xml,
    "yaml": parse_yaml,
}


class ParserOp(BaseOp):
    """Op that parses text into structured data.

    Supports multiple formats (JSON, XML, YAML) and extracts fields using
    dot-separated chain paths (e.g. ``"user.address.city: str"``). Commonly
    used as the final stage inside a ``ChainOp`` pipeline.

    Inputs:
        text (str): Raw text to parse (e.g. LLM output).

    Outputs:
        Dynamically generated from the ``extract`` list — one output key
        per extracted field.

    Example::

        parser = ParserOp(
            format="json",
            extract=["user.name: str", "user.age: int"],
            inputs={"text": llm["content"]},
        )
    """

    type: OpType = "parser"

    __slots__ = ["backend", "format", "extract", "extract_fields"]

    def __init__(
        self,
        format: ParserType = "xml",
        extract: Optional[List[str]] = None,
        inputs: Dict[str, Any] = None,
        outputs: Dict[str, Any] = None,
        **kwargs,
    ):
        if not extract:
            raise TypeError("extract là bắt buộc")

        # Parse schema thành format có cấu trúc
        extract_fields = [ExtractField.from_string(schema_str) for schema_str in extract]

        # Parse inputs/outputs từ extract
        parsed_inputs = {
            "text": Param(type=str, required=True),
            "validators": Param(type=dict, required=False),
        }
        parsed_outputs = {field.output_key: Param() for field in extract_fields}

        # Gọi super().__init__ không truyền inputs/outputs
        super().__init__(**kwargs)

        # Merge parsed schema with user-provided
        self._init_io(parsed_inputs, parsed_outputs, inputs, outputs)

        self.format = format or "xml"
        self.extract = extract
        self.extract_fields = extract_fields

        # Serialize parser config as literal inputs so Rust can read them
        mode_param = Param(type=str, required=False)
        mode_param.value = self.format
        self.inputs["mode"] = mode_param

        schema_param = Param(type=list, required=False)
        schema_param.value = self.extract
        self.inputs["schema"] = schema_param

        self.backend = self._create_parser()
        self._set_core(self._process)

    def _create_parser(self):
        """Tạo parser function dựa trên format."""
        # O(1) lookup for common formats
        parser = _PARSER_MAP.get(self.format)
        if parser is not None:
            return parser

        # Default fallback
        return parse_xml

    def _extract_value_by_path(self, data: Dict[str, Any], chain_path: List[str]) -> Any:
        """Trích xuất giá trị từ nested dictionary theo chain path."""
        current = data
        for key in chain_path:
            if isinstance(current, dict) and key in current:
                current = current[key]
            else:
                return None
        return current

    def _convert_type(self, value: Any, type_hint: str) -> Any:
        """Convert extracted value to the specified type.

        Handles type conversion for common type hints, especially boolean
        strings from XML parsing ("true"/"false" → True/False).

        Args:
            value: The raw extracted value
            type_hint: Type hint string (e.g., "bool", "int", "str", "float")

        Returns:
            Converted value, or original value if type hint is not recognized
        """
        type_hint = type_hint.lower().strip()

        # Boolean conversion - handles XML string "true"/"false"
        if type_hint in ("bool", "boolean"):
            if value is None:
                return None
            if isinstance(value, bool):
                return value
            if isinstance(value, str):
                value_lower = value.lower().strip()
                if value_lower in ("true", "1", "yes"):
                    return True
                if value_lower in ("false", "0", "no", ""):
                    return False
            return bool(value)

        # For non-boolean types, None stays None
        if value is None:
            return None

        # Numeric conversions
        if type_hint == "int":
            try:
                return int(value)
            except (ValueError, TypeError):
                return value

        if type_hint in ("float", "number"):
            try:
                return float(value)
            except (ValueError, TypeError):
                return value

        # String - convert and strip whitespace (XML often has \n around values)
        if type_hint in ("str", "string"):
            return str(value).strip() if value is not None else ""

        # For dict, list, any, or unknown types, return as-is
        return value

    async def _process(self, text: str, mode=None, schema=None, validators=None) -> Dict[str, Any]:
        """Parse text và trích xuất các field.

        Returns error as output (not exception) so downstream ops can check it.
        On success: {"field1": value, "field2": value, "error": None}
        On failure: {"field1": None, "field2": None, "error": "error message"}
        """
        LOGGER.debug(
            "ParserOp._process called: text=%r, validators=%r",
            text[:100] if text else text,
            validators,
        )
        if validators is not None and not isinstance(validators, dict):
            return {
                "error": f"validators must be a dict, got {type(validators).__name__}: {validators!r}"
            }
        if not text:
            return {"error": "Empty input text"}

        try:
            parsed_data = self.backend(text)
        except Exception as e:
            return {"error": f"Parse error ({self.format}): {e}"}

        result = {}
        for field in self.extract_fields:
            raw_value = self._extract_value_by_path(parsed_data, field.chain_path)
            result[field.output_key] = self._convert_type(raw_value, field.type_hint)

        # Validate extracted values against allowed lists (from input)
        # Values prefixed with @ are defaults (e.g. "@FALLBACK"):
        #   - @ is stripped for membership check (FALLBACK is a valid value)
        #   - When validation fails, the @-prefixed value is used as fallback
        if validators:
            for field_name, allowed_values in validators.items():
                clean_values = [v.lstrip("@") if isinstance(v, str) else v for v in allowed_values]
                default_value = next(
                    (
                        v.lstrip("@")
                        for v in allowed_values
                        if isinstance(v, str) and v.startswith("@")
                    ),
                    None,
                )
                value = result.get(field_name)
                if value is None or value not in clean_values:
                    if default_value is not None:
                        result[field_name] = default_value
                    else:
                        return {
                            "error": f"Validation failed: '{field_name}' value '{value}' not in {clean_values}"
                        }

        result["error"] = None
        return result

    @property
    def specific_metadata(self) -> Dict[str, Any]:
        """Trả về metadata riêng của subclass."""
        return {"format": self.format}

File: rust/operonx/src/core/ops/transform/parser_op.rs
//! `ParserOp` — extract structured data from JSON / XML / YAML text.
//!
//! Mirrors Python [`operonx/core/ops/transform/parser_op.py`](../../../../../operonx/core/ops/transform/parser_op.py).
//!
//! At op-runtime the executor reads four inputs off the resolved inputs map:
//!
//! | input        | type            | source                                   |
//! |--------------|-----------------|------------------------------------------|
//! | `text`       | string          | usually a ref to upstream LLM `content`  |
//! | `mode`       | string literal  | "json" / "xml" / "yaml" (default "xml")  |
//! | `schema`     | array literal   | list of `"a.b.c: type"` extract specs    |
//! | `validators` | object literal  | optional `{key: ["@DEFAULT", "v1", ...]}` |
//!
//! On success the executor returns `{<extracted_field>: <value>, ..., "error": null}`.
//! On failure it returns `{"error": "<msg>"}` so downstream ops can branch on it —
//! Python parity (see Python `_process` returning `{"error": ...}` instead of raising).

use async_trait::async_trait;
use quick_xml::events::Event;
use quick_xml::Reader;
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

use crate::core::exceptions::OperonError;
use crate::core::ops::base::{BaseOp, OpContext, OpMeta};

/// Parser format.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ParserType {
    Json,
    Xml,
    Yaml,
}

impl ParserType {
    pub fn from_str_lossy(s: &str) -> Self {
        match s.to_ascii_lowercase().as_str() {
            "json" => ParserType::Json,
            "yaml" | "yml" => ParserType::Yaml,
            _ => ParserType::Xml, // matches Python default
        }
    }
}

/// One extracted field — Python's `ExtractField.from_string("a.b.c: type")`.
#[derive(Debug, Clone)]
pub struct ExtractField {
    /// Last segment of the dot-path — the key the parsed value lands at.
    pub output_key: String,
    /// `["a", "b", "c"]` for `"a.b.c: int"`.
    pub chain_path: Vec<String>,
    /// `"int"` for `"a.b.c: int"`; `"Any"` if no `: T` suffix was given.
    pub type_hint: String,
}

impl ExtractField {
    /// Parse `"a.b.c: int"` (or `"a.b.c"` → type_hint "Any").
    pub fn from_string(schema_str: &str) -> Self {
        let (chain_text, type_hint) = match schema_str.find(':') {
            Some(i) => (&schema_str[..i], schema_str[i + 1..].trim().to_string()),
            None => (schema_str, "Any".to_string()),
        };
        let chain_path: Vec<String> = chain_text
            .trim()
            .split('.')
            .map(|s| s.to_string())
            .collect();
        let output_key = chain_path
            .last()
            .cloned()
            .unwrap_or_else(|| chain_text.to_string());
        ExtractField {
            output_key,
            chain_path,
            type_hint,
        }
    }
}

/// `ParserOp` — parse input text into structured data.
///
/// The struct itself is just metadata; the actual execution happens via the
/// free-function [`execute`] called from the scheduler's dispatch table.
pub struct ParserOp {
    pub meta: OpMeta,
    pub parse_as: ParserType,
    pub extract_fields: Vec<ExtractField>,
}

#[async_trait]
impl BaseOp for ParserOp {
    fn meta(&self) -> &OpMeta {
        &self.meta
    }

    async fn exec_core(
        &self,
        inputs: Map<String, Value>,
        _ctx: &OpContext<'_>,
    ) -> Result<Option<Value>, OperonError> {
        let val = execute(inputs).await?;
        Ok(Some(val))
    }
}

// ── Dispatch entry point — called from task_scheduler::execute_op ──────

/// Execute a parser op given a resolved inputs map. Returns a `Value::Object`
/// containing the extracted fields + an `error` key (null on success).
pub async fn execute(inputs: Map<String, Value>) -> Result<Value, OperonError> {
    // Mirrors Python `_process(text, mode, schema, validators)`:
    // - Validator type check returns {error: ...} (not exception).
    // - Empty text returns {error: "Empty input text"}.
    // - Parse failure returns {error: "Parse error (<mode>): <msg>"}.
    // - Validators with `@DEFAULT` semantics fall back when value not in allowed list.
    let text = inputs
        .get("text")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    let validators = inputs.get("validators").cloned();
    if let Some(v) = &validators {
        if !v.is_null() && !v.is_object() {
            return Ok(error_value(format!(
                "validators must be a dict, got {}: {:?}",
                json_type_name(v),
                v
            )));
        }
    }

    if text.is_empty() {
        return Ok(error_value("Empty input text".to_string()));
    }

    let mode = inputs
        .get("mode")
        .and_then(|v| v.as_str())
        .map(ParserType::from_str_lossy)
        .unwrap_or(ParserType::Xml);

    let parsed = match mode {
        ParserType::Json => parse_json(&text),
        ParserType::Xml => parse_xml(&text),
        ParserType::Yaml => parse_yaml(&text),
    };
    let parsed = match parsed {
        Ok(v) => v,
        Err(e) => {
            return Ok(error_value(format!(
                "Parse error ({}): {}",
                match mode {
                    ParserType::Json => "json",
                    ParserType::Xml => "xml",
                    ParserType::Yaml => "yaml",
                },
                e
            )))
        }
    };

    let fields = parse_extract_fields(inputs.get("schema"));

    let mut result = Map::new();
    for field in &fields {
        let raw = extract_value_by_path(&parsed, &field.chain_path);
        let converted = convert_type(raw, &field.type_hint);
        result.insert(field.output_key.clone(), converted);
    }

    // Validation pass — only when validators is a non-null object.
    if let Some(Value::Object(v_map)) = &validators {
        for (field_name, allowed) in v_map.iter() {
            let Some(allowed_arr) = allowed.as_array() else {
                continue;
            };
            let clean_values: Vec<Value> = allowed_arr
                .iter()
                .map(|v| match v.as_str() {
                    Some(s) if s.starts_with('@') => Value::String(s[1..].to_string()),
                    _ => v.clone(),
                })
                .collect();
            let default_value: Option<String> = allowed_arr.iter().find_map(|v| match v.as_str() {
                Some(s) if s.starts_with('@') => Some(s[1..].to_string()),
                _ => None,
            });
            let current = result.get(field_name);
            let pass = match current {
                Some(c) if !c.is_null() => clean_values.iter().any(|cv| values_equal(cv, c)),
                _ => false,
            };
            if !pass {
                if let Some(def) = default_value {
                    result.insert(field_name.clone(), Value::String(def));
                } else {
                    return Ok(error_value(format!(
                        "Validation failed: '{}' value '{}' not in {}",
                        field_name,
                        current.map(value_repr).unwrap_or_else(|| "None".into()),
                        list_repr(&clean_values)
                    )));
                }
            }
        }
    }

    result.insert("error".into(), Value::Null);
    Ok(Value::Object(result))
}

fn error_value(msg: String) -> Value {
    let mut m = Map::new();
    m.insert("error".into(), Value::String(msg));
    Value::Object(m)
}

fn json_type_name(v: &Value) -> &'static str {
    match v {
        Value::Null => "NoneType",
        Value::Bool(_) => "bool",
        Value::Number(_) => "number",
        Value::String(_) => "str",
        Value::Array(_) => "list",
        Value::Object(_) => "dict",
    }
}

fn parse_extract_fields(schema: Option<&Value>) -> Vec<ExtractField> {
    match schema.and_then(|v| v.as_array()) {
        Some(arr) => arr
            .iter()
            .filter_map(|v| v.as_str().map(ExtractField::from_string))
            .collect(),
        None => Vec::new(),
    }
}

fn extract_value_by_path(data: &Value, chain_path: &[String]) -> Value {
    let mut current = data.clone();
    for key in chain_path {
        match &current {
            Value::Object(m) => {
                if let Some(v) = m.get(key) {
                    current = v.clone();
                } else {
                    return Value::Null;
                }
            }
            _ => return Value::Null,
        }
    }
    current
}

/// Match Python `_convert_type` behavior on `Value`s.
fn convert_type(value: Value, type_hint: &str) -> Value {
    let hint = type_hint.trim().to_ascii_lowercase();
    match hint.as_str() {
        "bool" | "boolean" => match &value {
            Value::Null => Value::Null,
            Value::Bool(_) => value,
            Value::String(s) => {
                let l = s.trim().to_ascii_lowercase();
                if matches!(l.as_str(), "true" | "1" | "yes") {
                    Value::Bool(true)
                } else if matches!(l.as_str(), "false" | "0" | "no" | "") {
                    Value::Bool(false)
                } else {
                    Value::Bool(true) // matches Python `bool(value)` for non-empty
                }
            }
            other => Value::Bool(!is_falsy(other)),
        },
        _ if matches!(value, Value::Null) => Value::Null,
        "int" => match &value {
            Value::Number(n) => Value::Number(serde_json::Number::from(n.as_i64().unwrap_or(0))),
            Value::String(s) => s
                .parse::<i64>()
                .ok()
                .map(|n| Value::Number(n.into()))
                .unwrap_or(value),
            Value::Bool(b) => Value::Number(if *b { 1 } else { 0 }.into()),
            _ => value,
        },
        "float" | "number" => match &value {
            Value::Number(_) => value,
            Value::String(s) => s
                .parse::<f64>()
                .ok()
                .and_then(serde_json::Number::from_f64)
                .map(Value::Number)
                .unwrap_or(value),
            Value::Bool(b) => Value::Number(
                serde_json::Number::from_f64(if *b { 1.0 } else { 0.0 }).unwrap_or(0.into()),
            ),
            _ => value,
        },
        "str" | "string" => match value {
            Value::Null => Value::String(String::new()),
            Value::String(s) => Value::String(s.trim().to_string()),
            other => Value::String(value_repr(&other).trim().to_string()),
        },
        _ => value, // dict / list / Any / unknown — pass through
    }
}

fn is_falsy(v: &Value) -> bool {
    match v {
        Value::Null => true,
        Value::Bool(b) => !b,
        Value::Number(n) => n.as_f64().map(|f| f == 0.0).unwrap_or(false),
        Value::String(s) => s.is_empty(),
        Value::Array(a) => a.is_empty(),
        Value::Object(m) => m.is_empty(),
    }
}

fn values_equal(a: &Value, b: &Value) -> bool {
    match (a, b) {
        (Value::String(x), Value::String(y)) => x == y,
        (Value::Number(x), Value::Number(y)) => x
            .as_f64()
            .zip(y.as_f64())
            .map(|(a, b)| a == b)
            .unwrap_or(x == y),
        _ => a == b,
    }
}

fn value_repr(v: &Value) -> String {
    match v {
        Value::String(s) => s.clone(),
        other => other.to_string(),
    }
}

fn list_repr(items: &[Value]) -> String {
    let parts: Vec<String> = items
        .iter()
        .map(|v| match v {
            Value::String(s) => format!("'{}'", s),
            other => other.to_string(),
        })
        .collect();
    format!("[{}]", parts.join(", "))
}

// ── Format-specific parsers ────────────────────────────────────────────

fn strip_code_fences(text: &str) -> &str {
    let t = text.trim();
    if t.starts_with("```") {
        let mut lines: Vec<&str> = t.lines().collect();
        if lines.len() > 2 {
            // Drop the opening ```... and the closing ``` line.
            lines.remove(0);
            lines.pop();
            // Rejoining produces an owned String — return original for now and
            // handle in caller. Use the manual offset path instead.
            // Fall through to compute offsets:
        }
    }
    text
}

/// Parse JSON text. Strips code fences first to match Python's behavior.
fn parse_json(text: &str) -> Result<Value, String> {
    let cleaned = strip_fences_owned(text);
    serde_json::from_str(&cleaned).map_err(|e| e.to_string())
}

/// Parse YAML text. Strips code fences first.
fn parse_yaml(text: &str) -> Result<Value, String> {
    let cleaned = strip_fences_owned(text);
    let v: serde_yaml::Value = serde_yaml::from_str(&cleaned).map_err(|e| e.to_string())?;
    yaml_to_json(v).map_err(|e| e.to_string())
}

fn strip_fences_owned(text: &str) -> String {
    let t = text.trim();
    if t.starts_with("```") {
        let lines: Vec<&str> = t.lines().collect();
        if lines.len() > 2 {
            return lines[1..lines.len() - 1].join("\n");
        }
    }
    text.to_string()
}

fn yaml_to_json(value: serde_yaml::Value) -> Result<Value, String> {
    use serde_yaml::Value as Yv;
    match value {
        Yv::Null => Ok(Value::Null),
        Yv::Bool(b) => Ok(Value::Bool(b)),
        Yv::Number(n) => {
            if let Some(i) = n.as_i64() {
                Ok(Value::Number(i.into()))
            } else if let Some(f) = n.as_f64() {
                Ok(serde_json::Number::from_f64(f)
                    .map(Value::Number)
                    .unwrap_or(Value::Null))
            } else {
                Ok(Value::Null)
            }
        }
        Yv::String(s) => Ok(Value::String(s)),
        Yv::Sequence(seq) => {
            let v: Result<Vec<_>, _> = seq.into_iter().map(yaml_to_json).collect();
            Ok(Value::Array(v?))
        }
        Yv::Mapping(map) => {
            let mut out = Map::new();
            for (k, v) in map {
                let key = match k {
                    Yv::String(s) => s,
                    other => yaml_to_json(other)?
                        .as_str()
                        .map(String::from)
                        .unwrap_or_else(|| "".into()),
                };
                out.insert(key, yaml_to_json(v)?);
            }
            Ok(Value::Object(out))
        }
        Yv::Tagged(t) => yaml_to_json(t.value),
    }
}

/// Parse XML text using quick-xml. Mirrors Python's `xml_to_dict` shape:
///
/// - Element with no children: `{"tag": "text"}`.
/// - Element with children: `{"tag": {<child_tags>}}`.
/// - Repeated children with same tag: collapse to a list.
/// - Multiple roots: wrap in `<root>`, flatten as if children of `<root>`.
fn parse_xml(text: &str) -> Result<Value, String> {
    let cleaned = strip_fences_owned(text);
    // First try as-is.
    if let Ok(v) = parse_xml_inner(&cleaned) {
        return Ok(v);
    }
    // Multiple roots → wrap.
    let wrapped = format!("<root>{}</root>", cleaned);
    parse_xml_inner(&wrapped).map(|v| match v {
        Value::Object(mut m) => {
            if let Some(Value::Object(inner)) = m.remove("root") {
                Value::Object(inner)
            } else {
                Value::Object(m)
            }
        }
        other => other,
    })
}

fn parse_xml_inner(text: &str) -> Result<Value, String> {
    let mut reader = Reader::from_str(text);
    reader.config_mut().trim_text(true);

    // Stack of (tag_name, accumulator). On end-tag we collapse the
    // accumulator into either a string (no children) or an object.
    enum Frame {
        Empty,
        Text(String),
        Children(Map<String, Value>),
    }
    let mut stack: Vec<(String, Frame)> = Vec::new();
    let mut roots: Vec<(String, Value)> = Vec::new();

    loop {
        match reader.read_event() {
            Ok(Event::Start(e)) => {
                let name = std::str::from_utf8(e.name().as_ref())
                    .map_err(|e| e.to_string())?
                    .to_string();
                stack.push((name, Frame::Empty));
            }
            Ok(Event::End(_)) => {
                let (tag, frame) = stack
                    .pop()
                    .ok_or_else(|| "unbalanced XML end tag".to_string())?;
                let value = match frame {
                    Frame::Empty => Value::Null,
                    Frame::Text(s) => Value::String(s),
                    Frame::Children(m) => Value::Object(m),
                };
                // Merge into parent's Children frame, collapsing repeats to lists.
                if let Some((_p_tag, p_frame)) = stack.last_mut() {
                    let m = match p_frame {
                        Frame::Children(m) => m,
                        _ => {
                            *p_frame = Frame::Children(Map::new());
                            match p_frame {
                                Frame::Children(m) => m,
                                _ => unreachable!(),
                            }
                        }
                    };
                    if let Some(existing) = m.remove(&tag) {
                        match existing {
                            Value::Array(mut arr) => {
                                arr.push(value);
                                m.insert(tag, Value::Array(arr));
                            }
                            other => {
                                m.insert(tag, Value::Array(vec![other, value]));
                            }
                        }
                    } else {
                        m.insert(tag, value);
                    }
                } else {
                    roots.push((tag, value));
                }
            }
            Ok(Event::Text(t)) => {
                let s = t.unescape().map_err(|e| e.to_string())?.into_owned();
                if let Some((_tag, frame)) = stack.last_mut() {
                    if matches!(frame, Frame::Empty) {
                        *frame = Frame::Text(s);
                    } else if let Frame::Text(existing) = frame {
                        existing.push_str(&s);
                    }
                }
            }
            Ok(Event::Empty(e)) => {
                let name = std::str::from_utf8(e.name().as_ref())
                    .map_err(|e| e.to_string())?
                    .to_string();
                // Self-closing → represent as Null. Merge into parent.
                if let Some((_p_tag, p_frame)) = stack.last_mut() {
                    let m = match p_frame {
                        Frame::Children(m) => m,
                        _ => {
                            *p_frame = Frame::Children(Map::new());
                            match p_frame {
                                Frame::Children(m) => m,
                                _ => unreachable!(),
                            }
                        }
                    };
                    m.insert(name, Value::Null);
                } else {
                    roots.push((name, Value::Null));
                }
            }
            Ok(Event::Eof) => break,
            Ok(_) => {} // skip CData / Comment / Decl / PI / DocType
            Err(e) => {
                return Err(format!(
                    "XML parse error at pos {}: {}",
                    reader.error_position(),
                    e
                ))
            }
        }
    }

    if !stack.is_empty() {
        return Err("XML unbalanced — missing closing tag".to_string());
    }

    if roots.is_empty() {
        return Err("XML empty document".to_string());
    }

    if roots.len() == 1 {
        let (root_tag, root_value) = roots.pop().unwrap();
        let mut m = Map::new();
        m.insert(root_tag, root_value);
        Ok(Value::Object(m))
    } else {
        // Multiple roots — return as an object with all of them flattened.
        // Caller's wrapping path also handles this.
        let mut m = Map::new();
        for (tag, value) in roots {
            m.insert(tag, value);
        }
        Ok(Value::Object(m))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn extract_field_parses_dot_path_with_type() {
        let f = ExtractField::from_string("user.address.city: str");
        assert_eq!(f.output_key, "city");
        assert_eq!(f.chain_path, vec!["user", "address", "city"]);
        assert_eq!(f.type_hint, "str");
    }

    #[test]
    fn extract_field_defaults_to_any_when_no_type_suffix() {
        let f = ExtractField::from_string("user.name");
        assert_eq!(f.output_key, "name");
        assert_eq!(f.type_hint, "Any");
    }

    #[test]
    fn parse_json_strips_code_fence() {
        let v = parse_json("```json\n{\"a\": 1}\n```").expect("parse");
        assert_eq!(v, json!({"a": 1}));
    }

    #[test]
    fn parse_json_handles_raw_text() {
        let v = parse_json(r#"{"a": 1}"#).expect("parse");
        assert_eq!(v, json!({"a": 1}));
    }

    #[test]
    fn parse_xml_single_root() {
        let v = parse_xml("<root><name>Alice</name></root>").expect("parse");
        assert_eq!(v, json!({"root": {"name": "Alice"}}));
    }

    #[test]
    fn parse_xml_multiple_roots_get_flattened() {
        let v = parse_xml("<a>1</a><b>2</b>").expect("parse");
        // Per Python parity: wrapped in <root>, then xml_to_dict flattens.
        assert_eq!(v, json!({"a": "1", "b": "2"}));
    }

    #[test]
    fn parse_xml_repeated_tag_becomes_list() {
        let v = parse_xml("<root><item>1</item><item>2</item></root>").expect("parse");
        assert_eq!(v, json!({"root": {"item": ["1", "2"]}}));
    }

    #[test]
    fn parse_yaml_basic() {
        let v = parse_yaml("a: 1\nb: foo").expect("parse");
        assert_eq!(v, json!({"a": 1, "b": "foo"}));
    }

    #[test]
    fn convert_type_bool_true_strings() {
        assert_eq!(convert_type(json!("true"), "bool"), json!(true));
        assert_eq!(convert_type(json!("yes"), "bool"), json!(true));
        assert_eq!(convert_type(json!("1"), "bool"), json!(true));
    }

    #[test]
    fn convert_type_bool_false_strings() {
        assert_eq!(convert_type(json!("false"), "bool"), json!(false));
        assert_eq!(convert_type(json!(""), "bool"), json!(false));
        assert_eq!(convert_type(json!("0"), "bool"), json!(false));
    }

    #[test]
    fn convert_type_int_str_to_number() {
        assert_eq!(convert_type(json!("42"), "int"), json!(42));
    }

    #[test]
    fn convert_type_str_strips_whitespace() {
        assert_eq!(convert_type(json!("  hello  "), "str"), json!("hello"));
    }

    #[test]
    fn extract_value_by_path_walks_dict() {
        let data = json!({"a": {"b": {"c": 42}}});
        let v = extract_value_by_path(&data, &["a".into(), "b".into(), "c".into()]);
        assert_eq!(v, json!(42));
    }

    #[test]
    fn extract_value_by_path_missing_returns_null() {
        let data = json!({"a": 1});
        let v = extract_value_by_path(&data, &["a".into(), "b".into()]);
        assert_eq!(v, Value::Null);
    }
}
